text/XXX-flatten-feature.md

```diff
@@ -0,0 +1,325 @@
+- `--flatten --flatten-exclude=<buildpacks>` will flatten all the Buildpacks in one layer excepts the ones specified under the `flatten-exclude` flag 
+- `--flatten --flatten-include=<buildpacks>` will flatten the Buildpacks specified after the `flatten-include` flag into a single layer. Can be used more than once, with each use resulting in a single layer.
+
+We also need to define how a platform implementor needs to consume a flatten buildpackpage or builder.
+
+- When a platform consumes an OCI image artifact, they will need to inspect each buildpack layer blob and determine if the blob contains more than one buildpack, in such as case, they will need to process those buildpacks correctly.
+
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Let's say we have a composite buildpack (CB1) with the following dependency tree:
+```mermaid
+flowchart TD
+    A[CB1]
+    A --> B[G1]
+    A --> C[G2]
+    B --> BPA[BP1]
+    B --> BPFOO[BP2]
+    B --> BPC[BP4]
+    C --> BPD[BP1]
+    C --> BPBAR[BP3]
+    C --> BPE[BP4]
+```
+
+Until now, when a buildpack like this is being shipped  into an OCI image every individual buildpack is being saved in one layer, as a result we will have:
+
+$$ 
+ layer_1 = [CB_1] \\
+ layer_2 = [G_1] \\
+ layer_3 = [BP_1] \\
+ layer_4 = [BP_2] \\
+ layer_5 = [BP_4] \\
+ layer_6 = [G_2] \\
+ layer_7 = [BP_3] \\
+ total = \text{7 layers}
+$$
+
+Noticed that duplicated buildpacks are cleaned up. 
+
+We can use the new `flatten` flag to reduce the number of OCI image layers used by the buildpacks in different ways.
+
+* `--flatten`:
+Will flatten all buildpacks into one layer, the result will be:
+
+```mermaid
+classDiagram
+    class Layer1 {
+      CB1
+      G1 
+      BP1
+      BP2
+      BP4
+      G2
+      BP3
+    }
+```
+
+
+$$ 
+ total = \text{1 layer}
+$$
+
+---
+* `--flatten --flatten-exclude=<buildpack>` i.e. `--flatten-exclude=<BP2>`
+Will flatten all buildpacks in 1 layer except for`BP2`, the result will be
+
+```mermaid
+classDiagram
+    class Layer1 {
+      CB1
+      G1 
+      BP1
+      BP4
+      G2
+      BP3
+    }
+   class Layer2 {
+      BP2
+   }
+```
+
+$$ 
+ total = \text{2 layers}
+$$
+
+---
+
+* `--flatten --depth=0`
+Will flatten all layers at the root (0) level, which is similar to flatten all, the result will be:
+
+```mermaid
+classDiagram
+    class Layer1 {
+      CB1
+      G1 
+      BP1
+      BP2
+      BP4
+      G2
+      BP3
+    }
+```
+
+$$ 
+ total = \text{1 layer}
+$$
+
+---
+Increasing the depth number will allow to flatten the tree deeper,
+for example setting `--depth=1`, the result will be:
+
+```mermaid
+classDiagram
+    class Layer1 {
+      CB1
+    }
+    class Layer2 {
+      G1 
+      BP1
+      BP2
+      BP4
+    }
+    class Layer3 {
+      G2
+      BP3
+    }
+```
+
+
+$$ 
+ total = \text{3 layers}
+$$
+
+
+---
+
+* `--flatten --flatten-include=<buildpacks>` i.e. `--flatten --flatten-include=<BP1,BP2> --flatten-include=<BP3,BP4>`
```

Comment:
I like this for its simplicity and ease of understanding how buildpacks will be flattened together.

Reply:
We could potentially get rid of `--flatten-exclude` and even `--flatten` if the above is sufficient. We could perhaps introduce the special case of `--flatten-include *` if we want to flatten "all".

Reply:
We should validate it with our users just to be sure